Story/10956

kdwyer opened the pull request.

ws-mps commented on addons/edi_sale/models/edi_sale_request_document.py:
```diff
@@ -95,13 +95,32 @@ def prepare(self, doc):
+        # TODO check this query is efficient.
+        domain = [('id', 'in', orderless_partners.mapped('id')),
+                  ('sale_order_ids', '=', False)]
+        partners_with_no_historic_orders = Partner.search(domain)
```
Instead of a search, can we intersect `orderless_partners` with the partner records of the document?

kdwyer replied:
But wouldn't we end up trying to delete partners with empty orders in this import, but with orders from previous imports in the database?

ws-mps replied:
Yes, you are right.

ws-mps commented on addons/edi_sale/models/edi_sale_request_document.py:
```diff
@@ -95,13 +95,32 @@ def prepare(self, doc):
-            SaleRequestRecord = self.sale_request_record_model(doc)
-            reqs = SaleRequestRecord.search([('doc_id', '=', doc.id)])
-            for r, sales in reqs.mapped('sale_id').batched(self.BATCH_CONFIRM):
+            for r, sales in reqs.exists().mapped('sale_id').batched(self.BATCH_CONFIRM):
```
Is the `exists()` needed ? we are not deleting the sale order requests. 

Could we use `sales - to_remove` instead of `reqs.exists().mapped('sale_id')`?

kdwyer replied:
Done, please review.

ws-mps commented on addons/edi_sale/models/edi_sale_request_document.py:
```diff
@@ -95,13 +95,32 @@ def prepare(self, doc):
         super().execute(doc)
 
+        # Delete sale orders which no longer have any order lines.
+        SaleRequestRecord = self.sale_request_record_model(doc)
```
`SaleRequestRecord` and `Partner` should be in the same place, before or after `super()`.

kdwyer replied:
Done, please review.

kdwyer committed:
[IMP] edi: Add fail-fast flag to EDI Document

User-story/10956
Task/11085

Signed-off-by: Kevin Dwyer <kevin.dwyer@unipart.io>

To control whether document execution fails immediately when an error
occurs, or proceeds, skipping the failed line or order.

kdwyer committed:
[IMP] edi: Suppress missing relates if not in fail-fast mode.

User-story/10956
Task/11085

Signed-off-by: Kevin Dwyer <kevin.dwyer@unipart.io>

kdwyer committed:
[IMP] edi_sale: Skip invalid orders and order lines

User-story/10956
Task/11085

Signed-off-by: Kevin Dwyer <kevin.dwyer@unipart.io>

If we are not in fail-fast mode, do not process invalid order lines or
lines with no order (because we could not create the order).

Delete orders which are left with no lines.

Delete partners with no orders, as long as they do not have any historic
orders.

kdwyer committed:
[IMP] edi: Trap errors in synchronisation.

User-story/10956
Task/11086

Signed-off-by: Kevin Dwyer <kevin.dwyer@unipart.io>

If we are not in fail-fast mode, errors in order creation must not cause
document execution to terminate.

kdwyer committed:
[IMP] edi_sale: Create message listing unprocessed lines

User-story/10956
Task/11118

Signed-off-by: Kevin Dwyer <kevin.dwyer@unipart.io>

nocturnalastro commented on addons/edi/models/edi_synchronizer.py:
```diff
@@ -355,7 +356,17 @@ def execute(self):
+                        for vals in vals_list:
+                            try:
+                                instance = Target.create(vals)
+                            except ValidationError:
```
would it make sense to just put the if doc.fail_fast here? if so re-raise or something? Just thinking it minimizes the difference of each branch. Perhaps the comprehension is a little faster though,

kdwyer replied:
I did a crude `timeit` test, and on my machine the listcomp seems to be twice as fast as an accumulator loop (I was surprised the difference was so great).  So I think the listcomp gets to stay :smile:

nocturnalastro commented on addons/edi_sale/models/edi_sale_line_request_record.py:
```diff
@@ -69,12 +69,16 @@ def sale_line_values(self):
         # Identify containing EDI document
         doc = self.mapped('doc_id')
 
+        if not doc.fail_fast:
```
Not really bothered but can you put `doc.ensure_one()` before this so we don't get a cryptic error in the case we magically (in what ever strange scenario) get more than one doc. Also to signal to other people only expected one.

nocturnalastro commented on addons/edi_sale/models/edi_sale_request_document.py:
```diff
@@ -97,14 +97,47 @@ def execute(self, doc):
+        SaleLineRequestRecord = self.sale_line_request_record_model(doc)
+        lines = SaleLineRequestRecord.search([('doc_id', '=', doc.id)])
+        missings = lines.filtered(lambda l: not l.sale_line_id)
+        if not missings:
```
why not do if missing you do the same amount of work with only a single exit and it isn't overly nested anyway.

ws-mps replied:
Yes please:
``` 
if missing: 
    XXX

return
```

kdwyer replied:
Done, please review.

nocturnalastro commented on addons/edi_sale/models/edi_sale_request_document.py:
```diff
@@ -97,14 +97,47 @@ def execute(self, doc):
+    def _build_invalid_order_lines_report(self, doc, lines):
+        report_lines = ['Missing order lines']
+        for line in lines:
+            item = '\t'.join([str(x) for x in [line.order_key, line.product_key, int(line.qty)]])
```
perhaps make this a function to get the item that way if we want more info in the report we can overwrite it without touching the mechanism.

ws-mps replied:
I think this change is already on the way, good spot.

kdwyer replied:
I thought about that, but ended up thinking that the structure of the "report" is unspecified, so we might as well let subclasses overwrite the whole thing, and pass the resulting lines back to the caller.

kdwyer replied:
Apparently I've changed my mind :wink:

kdwyer replied:
Done, please review.

nocturnalastro commented on addons/edi_sale/tests/test_fail_fast.py:
```diff
@@ -0,0 +1,180 @@
+        doc.fail_fast = fail_fast
+        return doc
+
+    def assert_sales(self, sales):
```
`assert_sales_in_progress`?

ws-mps replied:
` def assert_sales_in_state(self, sales, state='sale'):` ?

kdwyer replied:
Done, please review.

nocturnalastro reviewed (approved):
The changes I've asked for are superficial so decided whether or not you want to do them.

nocturnalastro commented on addons/edi/models/edi_record.py:
```diff
@@ -110,6 +110,9 @@ class EdiRecord(models.AbstractModel):
                              required=True, readonly=True, index=True,
                              ondelete='cascade')
 
+    error = fields.Char(string='Error',
```
Why a string an not an edi.issue?

kdwyer committed:
[IMP] edi_sale: Clear errors after reporting them.

User-story/10956
Task/11137

Signed-off-by: Kevin Dwyer <kevin.dwyer@unipart.io>

Clear any recorded errors after they are reported so that we can
re-execute from a clean slate.
